File: source/kalk/dati/implement/stringfraz.cpp
#include "dati/define/stringfraz.h"

stringFraz::stringFraz(int a, int b, std::string c): frazione(a,b),mid(c){setName("stringFraz");}

stringFraz::stringFraz(const frazione& a, std::string b): frazione(a),mid(b){setName("stringFraz");}

stringFraz* stringFraz::clone() const{
    return new stringFraz(*this);
}

std::string stringFraz::getMid() const{
    return mid;
}

void stringFraz::setMid( std::string a){
    mid=a;
}

std::string stringFraz::toString()const{
    return std::to_string(getNum())+'/'+mid+'/'+std::to_string(getDen());
}

stringFraz* stringFraz::minima() const{
    frazione a=*this;
    a.reduce();
    return new stringFraz( a, mid);
}

stringFraz* stringFraz::complementare() const{
    frazione a=*this;
    if(razionale()>1)
        throw(std::string("stringFraz>=1 non puo avere complementare"));
    frazione* b=a.complementare();
    stringFraz* s=new stringFraz( *b, mid);
    delete b;
    return s;
}

stringFraz* stringFraz::operator +(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        frazione* f=frazione(*this)+a;
        stringFraz* s=0;
        if(frazione(*st)<this)
            s=new stringFraz( *f, st->mid);
        else
            s=new stringFraz( *f, mid);
        delete f;
        f=0;
        return s;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");

}

stringFraz* stringFraz::operator -(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        frazione* f=0;
        stringFraz* s=0;
        if(*st<this){
            f=frazione(*this)-st;
            s=new stringFraz( *f, mid);
        }
        else{
            f=frazione(*st)-this;
            s= new stringFraz( *f, st->mid);
        }
        delete f;
        f=0;
        return s;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");
}

stringFraz* stringFraz::operator *(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        frazione* f=frazione(*this) * st;
        stringFraz* s=0;
        if(*st<this)
            s=new stringFraz( *f, mid);
        else
            s=new stringFraz( *f, st->mid);
        delete f;
        f=0;
        return s;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");
}

stringFraz* stringFraz::operator /(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        frazione* f=0;
        stringFraz* s=0;
        if(*st<this){
            f=frazione(*this)/st;
            s=new stringFraz( *f, st->mid);
        }
        else{
            f=frazione(*st)/this;
            s=new stringFraz( *f, mid);
        }
        delete f;
        f=0;
        return s;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");
}

bool stringFraz::operator ==(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        return razionale()==st->razionale() && mid==st->mid;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");
}

bool stringFraz::operator !=(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        return razionale()!=st->razionale() && mid!=st->mid;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");

}

bool stringFraz::operator <(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        return razionale()<st->razionale() && mid<st->mid;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");
}

bool stringFraz::operator >(const frazione* a) const{
    if(const stringFraz* st=dynamic_cast<const stringFraz*>(a)){
        return razionale()>st->razionale() && mid>st->mid;
    }
    else
        throw("entrambi gli oggetti devono essere stringFraz");

}

File: source/java/stringFraz.java
import java.util.*;
public class stringFraz extends frazione{
    private String mid;
    public stringFraz(int n, int d, String s){
        super(n, d);
        mid=s;
        setName("stringFraz");
    }
    public void setMid(String s){
        mid=s;
    }
    public String getMid(){
        return mid;
    }@Override
    public stringFraz clone(){
        return new stringFraz(getNum(), getDen(), mid);
    }
    @Override
    public String toString(){
        return String.valueOf(getNum())+"/"+mid+"/"+String.valueOf(getDen());
    }
    @Override
    public stringFraz minima(){
        frazione a=super.clone();
        a.reduce();
        return new stringFraz(a.getNum(), a.getDen(), mid);
    }
    @Override
    public stringFraz complementare()throws alert{
        if(super.razionale()>1){
            throw new alert("frazione>1 non puo avere complementare");
        }
        frazione a=super.complementare();
        return new stringFraz(a.getNum(), a.getDen(), mid);
    }
    @Override
    public stringFraz somma(frazione f)throws alert{
        if(f instanceof stringFraz){
            frazione a=super.somma(f);
            if(super.maggiore(f))
                return new stringFraz(a.getNum(), a.getDen(), ((stringFraz)f).mid);
            else
                return new stringFraz(a.getNum(), a.getDen(), mid);
        }
        else 
            throw new alert("entrambi gli oggetti devono esere StringFraz");
    }    
    @Override
    public stringFraz differenza(frazione f)throws alert{
        if(f instanceof stringFraz){
            if(super.maggiore(f)){
                frazione a=super.differenza(f);
                return new stringFraz(a.getNum(), a.getDen(), mid);
            }
            else{
                frazione a= (frazione)f;
                a=a.differenza(this);
                return new stringFraz(a.getNum(), a.getDen(), ((stringFraz)f).mid);
            }
        }
        else 
            throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
    @Override
    public frazione moltiplicazione(frazione f)throws alert{
        if(f instanceof stringFraz){
            frazione a=super.moltiplicazione(f);
            if(super.maggiore(f))
                return new stringFraz(a.getNum(), a.getDen(), mid);
            else
                return new stringFraz(a.getNum(), a.getDen(), ((stringFraz)f).mid);
        }
        else 
            throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
    @Override
    public frazione divisione(frazione f)throws alert{
        if(f instanceof stringFraz){
            if(super.maggiore(f)){
                frazione a=super.divisione(f);
                return new stringFraz(a.getNum(), a.getDen(), ((stringFraz)f).mid);
            }
            else{
                frazione a= (frazione)f;
                a=a.divisione(this);
                return new stringFraz(a.getNum(), a.getDen(), mid);
            }
        }
        else 
            throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
    @Override
    public boolean ugualianza(frazione f)throws alert{
        if(f instanceof stringFraz)
            return razionale()==f.razionale() && mid==((stringFraz)f).mid;
        throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
    @Override
    public boolean diverso(frazione f)throws alert{
        if(f instanceof stringFraz)
            return razionale()!=f.razionale() && mid!=((stringFraz)f).mid;
        throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
    @Override
    public boolean maggiore(frazione f)throws alert{
        if(f instanceof stringFraz)
            return razionale()>f.razionale() /*&& mid==((stringFraz)f).mid*/;
        throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
    @Override
    public boolean minore(frazione f)throws alert{
        if(f instanceof stringFraz)
            return razionale()<f.razionale() /*&& mid==((stringFraz)f).mid*/;
        throw new alert("entrambi gli oggetti devono esere StringFraz");
    }
}